### ISSUE TITLE
Idea/Proposal: Allow for public libraries to depend on private libraries

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#1017]
## Abstract

Currently, private libs may depend on public libs but the reverse isn't allowed. Note that an analogous situation exists with executables, but this restriction is absent. This proposal seeks to harmonize the situation between public libs and public exes. Both should be allowed to consume private libs as dependencies.

### Issues to address

To solve this problem, three issues must be addressed:

* Private libs should work the same way as public libs. Regardless of whether they are consumed from findlib or from a local workspace, the handling of deps should be the same.

* Private libraries should be hidden from users of the public libs

* There should be a guarantee that public libs do not leak the types of private libraries in their interface.

### Proposed solution

_tl;dr_ is we'll install the private libs alongside the public libs in findlib and differentiate the two with a naming convention. Private libs will be guarded from downstream users by hiding the cmi's. To guard private types escaping into public interfaces, we won't allow the user to compile the public interfaces against the private lib's cmi's.

The first issue is really an issue when the public libs (with private deps) are installed as findlib libs. The private libs must also be installed alongside. This can be accomplished with some naming convention for private libs. Note that we cannot just "inline" the object files of a private lib with the public lib because multiple public libs can share the same private libs. I understand this is a huge hack, though hopefully real namespaces would allow for a simpler solution.

The second issue is also resolved by not installing the .cmi's of the private libs. This is the same mechanism that we'll use for implementing private modules.

The last issue is resolved by making sure that the interfaces of public libs will not be compiled against the interfaces of private libs. Hence, we'll exclude paths of the .cmi's of private deps when building the .cmi's of the public libs. This should make sure that no types defined in a private lib can end up in the interface of a public lib.

### Possible issues

There's a bit of a complication with private modules. Private modules should really be able to export the types of private libraries.

**Comment by nojb** [comment]
> There should be a guarantee that public libs do not leak the types of private libraries in their interface.

Isn't this too strong? In my opinion, "private" should mean ".cmi not installed". This still allows to expose types from private libraries/modules in the public interface. In our experience it is useful to do this. Let me explain what I mean in a toy explample.

Suppose `priv.ml` contains
```ocaml
type t = int
let x : t = 1
```
`pub.mli` contains
```ocaml
val x : Priv.t
val f : Priv.t -> unit
```
and `pub.ml`
```ocaml
let x = Priv.x
let f = print_int
```
Then we can compile modules in our "library" with
```
ocamlc -c priv.ml pub.mli pub.ml
```
but afterwards we get rid of `priv.cmi`:
```
rm priv.cmi
```
Now suppose we have `main.ml`
```ocaml
let () = Pub.f Pub.x
```
Then we can still compile
```
ocamlc priv.cmo pub.cmo main.ml
```
but in this compilation `Priv.t` is **abstract**. Note that even if you cannot write `Priv.t` in `Main` (because the `.cmi` is not available), you could have re-exported it from `Pub` by adding `type t = Priv.t` in `pub.mli` and then you could use the abstract manifest type `Pub.t = Priv.t` in `Main`.

At LexiFi we ship a plugin editor that allows to write OCaml plugins against a public OCaml API and we enforce this by using the above method. Not being able to expose "private" types as abstract would be too restrictive.

So, in a nutshell, I think **not** requiring that types from private libraries not be exposed in public libraries will simplify implementation and make it more powerful.

**Comment by rgrinberg** [comment]
> At LexiFi we ship a plugin editor that allows to write OCaml plugins against a public OCaml API and we enforce this by using the above method. Not being able to expose "private" types as abstract would be too restrictive.

I see. I would have thought that explicitly re-exporting the private types as abstract in the public library would be the recommended practice here, but maybe this is unwelcome boilerplate.

I still think that having a mode that prevents you from leaking private types is useful, but I tend to agree with you that it's probably an unnecessary restriction.

**Comment by diml** [comment]
Note that it would also be hard to enforce this property.

**Comment by rgrinberg** [comment]
Why is it hard? Isn't it just a matter of not including the private lib's cmi's when compiling public libs mlis? The only annoyance is that the error messages aren't going to be so good.

**Comment by diml** [comment]
That's true, I was thinking of enforcing it afterwards. FTR, when I used private modules in the past, I was also referring to types from private modules in public mli files.

**Comment by jaredly** [comment]
tbh I think it makes sense to require that types from private libraries be re-exported

**Comment by bobot** [comment]
Which public libraries could depend on a private libraries?

I believe only the one of the same package because compilation and installation of different package could be separated with `-p`. Or library of the same scope if some combination of `-p` are mandatory.

**Comment by SaswatPadhi** [comment]
Hello,

For our project, we are facing this issue -- we are creating subdirectories with `dune` files without a `public_name` for the library, so those are sort of like private libraries. We would like to use these private libraries in the `dune` file in the parent directory which builds a public library.
Does the current version of `dune` provide another way of achieving similar functionality?

Thanks.

**Comment by Lupus** [comment]
Would be awesome to have such feature supported by dune! We currently have some libraries that are built from a bunch of sub-libraries. Naively naming private sub-libraries something like "lib" we later ran into issues that "lib" ends up being top level library when original library is linked in.

It's sometimes convenient to have unwrapped library, for example to bundle all atdgen specs, link this lib and have all of atdgen produced modules available, but as this lib is used by public library, it also has to be public, and it pollutes the namespace with those atdgen modules.

We'll have to mangle private library names ourselves manually to somewhat hide them, and have atdgen bundle lib be wrapped. Would be cool if dune could do that for us.

**Comment by yawaramin** [comment]
@Lupus interesting that you ran into this ... I did too recently and sent a doc PR #3111 to warn about it. I think the Dune folks want to add a config option to solve the nested libs issue.

**Comment by rgrinberg** [comment]
> but as this lib is used by public library, it also has to be public, and it pollutes the namespace with those atdgen modules.

Note that you will still need to attach this library to a package. Otherwise you have two libraries in different packages that use the private library. In this case, which package should install the private lib?

**Comment by diml** [comment]
BTW, from the other discussion and this one, something is not clear to me. Are you interested in:

1. the ability to have a public library depend on a private library
2. the ability to have a library be split across several directories, with the directory names  reflected in the module system. i.e. `foo/bar.ml` seen as `Bar` in `foo/x.ml` and as `Foo.Bar` in `y.ml`

?

**Comment by Lupus** [comment]
> Note that you will still need to attach this library to a package. Otherwise you have two libraries in different packages that use the private library. In this case, which package should install the private lib?

That's not a problem, all of the to-be-private sub-libraries are in one package. And in general it makes sense that public library from a package can depend only on private libraries from the same package.

> the ability to have a library be split across several directories, with the directory names reflected in the module system. i.e. foo/bar.ml seen as Bar in foo/x.ml and as Foo.Bar in y.ml

The above was actually my goal. Somewhere in dune documentation (if I'm not mistaken) there was a recommendation to make more libraries to organize the codebase as creating those is cheap. It indeed helps in organizing large collections of files into clusters by their relevance. Having the ability to split library into a bunch of subfolders such that hierarchy gets reflected in module paths also solves this, especially if top level folder could provide a .mli/.rei for the whole library to filter out what's not supposed to be public.

**Comment by rgrinberg** [comment]
Check out the (include_subdirs qualified) proposal then. We’re actually not too far from implementing anymore as all the prerequisite refactoring has been done.

**Comment by Lupus** [comment]
Just checked https://github.com/ocaml/dune/issues/1084, looks awesome! Given that module organization within one library can be solved with that, just curious, what would be the use case for depending on a private library in a public one? Share some code between executable and public library within one project?

**Comment by rgrinberg** [comment]
That’s one use case. Another one is when you can partition a library so that they have different dependencies. Think foo and foo_lwt for example. In this case, it’s likely that you’ll want to share some code between these two libraries but maintain them in separate packages.

**Comment by emillon** [comment]
I was about to write a proposal for this, but I had forgotten about this issue... Anyway here's my suggestion:

Sub-libraries that start with an underscore character, such as `(public_name a._b)` get special treatment by dune:

- they can only appear in the `(libraries)` field of `a`, its sub-libraries (public or private), or private executables.
- its cmi files are not installed

I believe that this is enough for many use cases, such as wanting to test internals of a library and defining a public API at the same time.

If this works, we can change how findlib deals with this as well, so that for example `ocamlfind list` does not show these private sublibraries.

**Comment by rgrinberg** [comment]
Here’s a simpler version of this proposal that was discussed before:

There’s no need to add a public_name for such libraries. They can be referred by their private names and we can specify the package using the package field like we do for tests, executables.

**Comment by jeremiedimino** [comment]
@rgrinberg, with this proposal, what happens in this case:

- public library `p1.a` depends on private library `x`
- public library `p1.b` depends on private library `x`
- executable `prog` that is part of another `p2` package depends on both `p1.a` and `p1.b`

How do we ensure that `x` is linked only once in `prog`?

**Comment by rgrinberg** [comment]
My proposal still requires we specify the package of x. So x will be installed as part of that package.

**Comment by jeremiedimino** [comment]
Indeed. In which directory would such a private library be installed?

**Comment by rgrinberg** [comment]
> Indeed. In which directory would such a private library be installed?

We could use a naming convention like suggester earlier. A private library `foo` in a package `bar` would be `foo._bar`. I'm not even sure if we need the obfuscation even.

**Comment by jeremiedimino** [comment]
I see, so it's really just a different presentation. Yes, that makes sense to me and I indeed like it better, this way the user doesn't have to come up with a public name.

Regarding the obfuscation, if we don't do it we might have clashes. For instance If a public library is already called `foo.bar`. We also don't want such libraries to appear in the listing when doing `ocamlfind list` or `dune installed-libraries`.

**Comment by rgrinberg** [comment]
> Regarding the obfuscation, if we don't do it we might have clashes. For instance If a public library is already called foo.bar

That’s right, but I was hoping the underscore would help prevent that. So we’d have foo._bar for the private library. That’s still not bullet proof, but I think we do need a convention of sorts to solve the next problem:
> We also don't want such libraries to appear in the listing when doing ocamlfind list or dune installed-libraries.

ocamlfind and dune would both respect the naming convention above.

Another possible naming convention is to reserve the `foo.private` prefix for private libraries. So, `foo.private.*` would all be private libraries. That seems a little more obvious to me.

**Comment by Lupus** [comment]
> Regarding the obfuscation, if we don't do it we might have clashes.

+1 for obfuscation! We tend to have atd specifications piled into separate library because dune flags are different for those (does not work with `-open Base`), and just name those libs `atd_specs`. When some public library depends on such "sub-library", we have to manually prefix the private name with public name so that it does not clash with `atd_specs` from another library... That results in a bunch of ugly module opens/aliases on top like `module Atd_specs = My_public_lib__atd_specs`.

**Comment by jeremiedimino** [comment]
> Another possible naming convention is to reserve the `foo.private` prefix for private libraries. So, `foo.private.*` would all be private libraries. That seems a little more obvious to me.

What about `foo.__private__`? We are pretty sure this won't clash with other names and it seems even more private.

@Lupus we were not talking about this kind of obfuscation here, though we could imagine it. In any case, it seems that what you really want is the ability to specify flags per file, which is another feature we have been talking about adding and should happen at some point.

**Comment by Lupus** [comment]
Oh, sorry for confusion. Flags per file sound nice, in tandem with multi-folder libraries probably should be sufficient to structure large library codebases without splitting in separate libs due to artificial reasons like compilation flags.

**Comment by jeremiedimino** [comment]
Just to sum up, the proposal becomes:

the user writes:

```
(library
 (name foo)
 (package p))
```

and it gets installed as `p.__private__.foo` without the cmi files. @emillon what do you think?

**Comment by emillon** [comment]
That seems good, yes.

I'm not 100% sure about the `__private__` since the name is already overloaded with different terms and it'd be nice to have something that can be easily searched for if somebody discovers this kind of libraries on their system. But the semantics are in line with what I was suggesting, yes!

**Comment by rgrinberg** [comment]
Ready in master :)
